cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg/persistence/Common.kt

```diff
@@ -0,0 +1,244 @@
+        if (labelCache.containsKey(cacheKey)) {
+            return labelCache[cacheKey] ?: setOf()
+        }
+
+        val labels = mutableSetOf<String>()
+        labels.addAll(this.superclasses.flatMap { it.labels })
+        this.simpleName?.let { labels.add(it) }
+
+        // update the cache
+        labelCache[cacheKey] = labels
+        return labels
+    }
+
+/** A list of specific types that are intended to be ignored for persistence. */
+internal val ignoredTypes =
+    listOf(
+        TranslationContext::class.createType(),
+        TranslationConfiguration::class.createType(),
+        BenchmarkResults::class.createType(),
+        KClass::class.createType(listOf(KTypeProjection.STAR)),
+    )
+
+internal val nodeType = Node::class.createType()
+internal val collectionType = Collection::class.createType(listOf(KTypeProjection.STAR))
+internal val mapType = Map::class.createType(listOf(KTypeProjection.STAR, KTypeProjection.STAR))
+
+/**
+ * Retrieves a map of schema properties (not relationships!) for the given class implementing
+ * [Persistable].
+ *
+ * This property computes a map that associates property names (as strings) to their corresponding
+ * [KProperty1] objects, which represent the properties defined in the class. Only properties whose
+ * return types are included in a predefined set of supported property types ([propertyTypes]) are
+ * included in the map.
+ *
+ * The computed map is cached to optimize subsequent lookups for properties of the same class.
+ */
+val KClass<out Persistable>.schemaProperties: Map<String, KProperty1<out Persistable, *>>
+    get() {
+        // Check, if we already computed the properties for this node's class
+        return schemaPropertiesCache.computeIfAbsent(this) {
+            val schema = mutableMapOf<String, KProperty1<out Persistable, *>>()
+            val properties = it.memberProperties
+            for (property in properties) {
+                if (isSimpleProperty(property)) {
+                    schema.put(property.name, property)
+                }
+            }
+            schema
+        }
+    }
+
+/**
+ * Evaluates whether a given property qualifies as a "simple" property based on its characteristics.
+ *
+ * This evaluates to false, when
+ * - The property is a list (see [collectionType])
+ * - The property is a map (see [mapType])
+ * - The property is one of our [ignoredTypes]
+ * - The property is referring to a [Node]
+ * - The property is an interface
+ *
+ * @param property the property to be evaluated, belonging to a class implementing the Persistable
+ *   interface
+ * @return true if the property satisfies the conditions of being a "simple" property, false
+ *   otherwise
+ */
+private fun isSimpleProperty(property: KProperty1<out Persistable, *>): Boolean {
+    val returnType = property.returnType.withNullability(false)
+
+    return when {
+        property.visibility == KVisibility.PRIVATE -> false
```

Comment:
I am not sure if we want to have the visibility private to be a reason for not persisting a property and if it is a readoson it should definitely be its own function `isPrivate()`

Reply:
We can't access private properties without major reflection foo, that's why I chose to ignore them.